### ISSUE TITLE
Client: saner handling of fields value changes

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#1732]
`xforms.js` and `AjaxServer.js` use `ORBEON.xforms.Globals.changedIdsRequest` to store information about `keyup` and `keydown` events in input and password fields and in textareas.

The code doesn't make it very clear, to say the least, what the purpose of it is. Here is what I managed to figure out:
- `changedIdsRequest` is used as a hashmap
  - `String → Int`
  - key: control id
  - value: number of pending `keydown` events in the field without a `keyup`
    - so value is `undefined` or `null` when cleared
    - or value is at `0` once a change has made it to the field
- a value in the map can be `undefined`, ~~`null`,~~ or `Int`
- ignore keycodes which don't modify the field (`undefined`, `9`, `16`, `17`, `18`)
- ignore all fields except `xforms-input`, `xforms-secret`, and `xforms-textarea`
- `keydown` for non-ignored keycodes for those fields
  - for control id, add value `1` in map if no key
  - else increment value by `1`
- `keyup` for non-ignored keycodes for those fields
  - decrement value by 1
- `change` (tab out of field typically)
  - if there is value, set it to `0`
  - says don't receive `keyup` when that happens, could this be for the tab key only?
- if there is a value in the map for a control, `setCurrentValue` has no effect on that control
- `AjaxServer.js`
  - `executeNextRequest`: just before sending events to server, set to `null` all values which are `0`
  - `handleResponseAjax`: after `handleResponseDom`, clear map entirely (set to `{}`)

**Comment by ebruchez** [comment]
Seeing this again. It is puzzling. I moved the code to Scala to get clarity. Confirming the following:

- `setCurrentValue()` uses `changedIdsRequest` to ignore actually setting the value in some cases
- just before sending events to the server, `changedIdsRequest` is pruned of all entries which have a value of `0`
- `changedIdsRequest` is completely cleared just after processing a server response
- All the other usages either set, increment, or decrement the values.

**Comment by ebruchez** [comment]
Moving to a separate file. I thought this might have been an issue with #4503, but it appeared completely independent.

My understanding so far is that the only *output* of this entire process, with `hasChangedIdsRequest`, is whether a control's value "has changed since we sent our request".\

This is used when `setCurrentValue()` is called, and for input/textarea/select/maybe other controls. This is why a missing value in the map is not the same as a value set to `0`.
 
`setCurrentValue()` is called:

- by `DocumentAPI.setValue`
- and upon processing the Ajax response

If a control has `hasChangedIdsRequest == true`, then setting the value doesn't happen (except when a `force` attribute is passed!).

So the purpose of this appears to be simply "when processing an Ajax response, don't override values modified by the user since the request was sent", which rings a bell.

Typically, this would apply to a field the user is currently typing in, making changes, and then the server would respond and overwrite the value. We don't want this to happen.

Now:

- Is this implemented correctly? Probably not!
- `DocumentAPI.setValue` should probably not go through this

**Comment by ebruchez** [comment]
Can we express this logic in a simple way? This must involve:

- keyboard events within input/textarea or some other way of telling us the value has changed "since"
- maybe custom API for XBL components like rich text
- the timing of Ajax requests/responses

Scenario:

- I am in a field 1, change a value, tab to field 2
- request goes to server 
- server modifies value of field 2
- in the meanwhile, I am typing in field 2
- response arrives with new value for field 2
- value from server does not modify field 2
- as usual, value from field 2 will be sent to server either after incremental delay or when tabbing out

Do we need to handle fields changed but that we have left? I.e. same as above, but while the Ajax request is in progress, I have made changes in field 2, moved back to field 1, made changes there, and the server wants to update field 1 and field 2?

The fields I have left have queued events in the event queue, ready to be sent back to the server. While I am not in field 1 or field 2, the server would update the field I am not in, before the value is sent to the server. But the server would not tell us to update the field again, so the field would stay with the old value.

In other words, it seems that we cannot *just* handle the current field: we must handle all the modified fields.

We cannot just look at the event queue, because that only contains the values for fields that have changed (either upon `keyup` but only for incremental fields, or when tabbing out for non-incremental fields).

Now what is the boundary where the list of modified fields must be set?

**Comment by ebruchez** [comment]
Scenario that might not work right now:

- user types in non-incremental field but doesn't tab out
- due to for example  a timeout, Ajax request is scheduled
    - list of modified fields is cleared
- server modifies value of field
- field is overwritten

This issue can only happen with a current non-incremental field.

What about a slightly different algorithm:

- keep list of modified fields
    - includes value changes in the event queue
    - AND changes to the unique "current" field that have not yet translated to a change in the event queue
- just before *sending* events to the server
    - clear the list of modified fields except for the "current" field (if its value is different from any in the queue?)
- when a response arrives
    - don't call `setCurrentValue()` when processing the response for values in the current list of changed values

With this:

- We only need to keep a single "current field is being modified"
- we can check the event queue for the rest
- we cover more uses cases

Q: Do we still need the `keyup`/`keydown` incrementing/checking? 

Philosophy: "Any field with a value change initiated by the user which has not yet been sent to the server (giving the server a chance to process the value change) will ignore changes specified by the server."

We know that these user-initiated changes will eventually made it to the server, as they either:

- are in the event queue already
- will be in the event queue once the user tabs out

(The user may never tab out but then arguably either it's ok or something else should cause that value to be committed, see also #4444).

**Comment by avernet** [comment]
"Any field with a value change initiated by the user which has not yet been sent to the server (giving the server a chance to process the value change) will ignore changes specified by the server." Yes, and I thought that this is what we had implemented.

**Comment by ebruchez** [comment]
Yes, it's what we implemented (minus some possible bugs, as mentioned above), but I could only make sure by attempting to reverse-engineer the code ;)

**Comment by ebruchez** [comment]
There is still one open question about the meaning of the "increment then decrement value and clear non-zero values" logic.

Before sending an Ajax request, we clear all changes in the map, *except* those that are not at `0`. Those at `0` are the ones that got matching `keydown`/`keyup` events. I guess the intent was "some `keydown` arrived but no `keyup` yet". But also, upon `change`, we set the value to `0`. Clearing the map means that we consider that all the values are about to be sent to the server. However, this is not guaranteed (see timeout scenario about). But it works in the incremental case, and when tabbing out.

This said, with Chrome, Safari, and Firefox on Mac, I noticed that holding down a key (which shows the selector for accented letters) keeps sending repeated `keydown` events and we just get one corresponding `keyup`! So the counting is broken in this case. It will fail for an incremental field,

The decrementing might also reach `-1` depending on when exactly we call `processEvents`.

My take: this incrementing/decrementing logic is currently very broken.

I note that browsers have an [`input`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event) event, which " is fired every time the value of the element changes. This is unlike the change event, which only fires when the value is committed, such as by pressing the enter key, selecting a value from a list of options, and the like." Could we use this instead? There is also a [`beforeinput`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/beforeinput_event) but it's not widely implemented yet.

**Comment by ebruchez** [comment]
Quick test of the `input` event: it appears to work with `<input>` and `<textarea>` with all modern browsers. IE 11 also dispatches it, even for `<textarea>` (while MDN was unclear about that). So I think we can use this event. No need for complex `keydown`/`keyup` (incorrect) counting and checking.

**Comment by ebruchez** [comment]
Talking with @avernet:

1. We agree that the current behavior is not correct.
2. We need scenarios.

Scenario 1:

- field has server-side formatting, for example formatting a number or value
- user enters value in non-incremental field, then tabs out 
- request is scheduled to be sent and eventually is sent
- in the meanwhile, user tabs back to field and modifies value
- servers responds with formatted value for previous value
- field should not update

- [ ] make sure this scenario works as intended
- [ ] check behavior of `fr:number`
